src/astro/sql/table.py

```diff
@@ -79,6 +76,42 @@ def sqlalchemy_metadata(self) -> MetaData:
             alchemy_metadata = MetaData()
         return alchemy_metadata
 
+    def __getattribute__(self, attribute_name):
```

Comment:
Can we not create a `name` property it as follows, similar to the example in https://docs.python.org/3/library/functions.html#property:

```python
@property
def name(self):
     # Custom logic you have in getter 
     if self.temp and not self._name:
         return self._create_unique_table_name()
     return self._name
```

Reply:
@kaxil, my first attempts used properties, and they were not successful. Some attempts I tried:

**Attempt 1**

```
class Table:
    _name: str=""

    @property
    def name(self) -> str:
        if self.temp and not self._name:
            return self._create_unique_table_name()

    @name.setter
    def name(self, value: str) -> None:
        self._name = value
```

Which resulted in all tests failing:
```
    def test_table_with_explicit_name():
        """Check that we respect the name of table when it's named an we set it to non-temp"""
        table = Table(conn_id="some_connection", name="some_name")
        assert table.conn_id == "some_connection"
>       assert table.name == "some_name"
E       AssertionError: assert None == 'some_name'
E        +  where None = Table(conn_id='some_connection', name=None, metadata=Metadata(schema=None, account=None, database=None, host=None, region=None, role=None, warehouse=None), columns=[], temp=False).name

tests/sql/test_table.py:8: AssertionError
------------------------------------------------------------------------------- Captured stdout setup --------------------------------------------------------------------------------
[2022-05-20 14:16:23,654] {crypto.py:82} WARNING - empty cryptography key - values will not be stored encrypted.
--------------------------------------------------------------------------------- Captured log setup ---------------------------------------------------------------------------------
WARNING  airflow.models.crypto:crypto.py:82 empty cryptography key - values will not be stored encrypted.
______________________________________________________________________________ test_table_without_name _______________________________________________________________________________

    def test_table_without_name():
        """Check that we create a name, that it is always the same name, and that we set temp name to True"""
        table = Table(conn_id="some_connection")
        assert table.conn_id == "some_connection"
        name1 = table.name
        name2 = table.name
>       assert name1
E       assert None

tests/sql/test_table.py:18: AssertionError
_________________________________________________________________________ test_table_without_name_and_schema _________________________________________________________________________

    def test_table_without_name_and_schema():
        """Check that the table name is smaller when there is metadata associated to the table."""
        table = Table(conn_id="some_connection")
        table.metadata.schema = "abc"
>       assert isinstance(table.name, str)
E       AssertionError: assert False
E        +  where False = isinstance(None, str)
E        +    where None = Table(conn_id='some_connection', name=None, metadata=Metadata(schema='abc', account=None, database=None, host=None, region=None, role=None, warehouse=None), columns=[], temp=False).name

```

**Attempt 2**

```
    @property
    def name(self) -> str:
        if self.temp and not self._name:
            return self._create_unique_table_name()

    @name.setter
    def name(self, value: str) -> None:
        if type(value) is not property:
            self._name = value

```

Which resulted in:
```
    def test_table_with_explicit_name():
        """Check that we respect the name of table when it's named an we set it to non-temp"""
        table = Table(conn_id="some_connection", name="some_name")
        assert table.conn_id == "some_connection"
>       assert table.name == "some_name"
E       AssertionError: assert None == 'some_name'
E        +  where None = Table(conn_id='some_connection', name=None, metadata=Metadata(schema=None, account=None, database=None, host=None, region=None, role=None, warehouse=None), columns=[], temp=False).name

tests/sql/test_table.py:8: AssertionError
------------------------------------------------------------------------------- Captured stdout setup --------------------------------------------------------------------------------
[2022-05-20 14:18:58,617] {crypto.py:82} WARNING - empty cryptography key - values will not be stored encrypted.
--------------------------------------------------------------------------------- Captured log setup ---------------------------------------------------------------------------------
WARNING  airflow.models.crypto:crypto.py:82 empty cryptography key - values will not be stored encrypted.
______________________________________________________________________________ test_table_without_name _______________________________________________________________________________

    def test_table_without_name():
        """Check that we create a name, that it is always the same name, and that we set temp name to True"""
        table = Table(conn_id="some_connection")
        assert table.conn_id == "some_connection"
        name1 = table.name
        name2 = table.name
        assert name1
        assert isinstance(name1, str)
        assert len(table.name) == 62
>       assert name1 == name2
E       AssertionError: assert 'o7dkhq82gf0a...z3lt5lieag5to' == 'y91dzxtv8s7h...y2ilklpivkum6'
E         - y91dzxtv8s7hm231hnobldu8wquc6z8cgruknp2o45gap3q0gy2ilklpivkum6
E         + o7dkhq82gf0am1k2f1mt9gpg3iu7xmnvkwsjkc10xvd0w9jniz3lt5lieag5to

tests/sql/test_table.py:21: AssertionError

```

I made several attempts of using properties changing the field types and the data class definition itself. Since we have other priorities, I thought it was better to have a working ugly version and move to the other more critical PRs - we can replace the current implementation or an alternative working one!

Reply:
@tatiana in the property code you wrote, it looks like you don't return a name if temp=false. Could that be why tests were failing?

Reply:
Great catch, @dimberman ! Thank you :heart:  I just updated the code and pushed a new version :tada: